### PR TITLE
fix RotOverhead camera rotating wrong way after reaching half rotation

### LINKED ISSUE #16: Something is wrong with wrap-around for FPS/rot camera
`Spring.SetCameraState(Spring.GetCameraState(), cameraTransitionTime)`

causes this issue.

spring camera is immune.

### PR DESCRIPTION
**Pull request by dzosz**
Draft. I still need to understand few more things to make sure if this is right approach.
1. Why Free camera is not affected by this problem? Edit: Free camera does not preserve dir/rotation and always reflects camera position so it is never affected by transition mixing. 
2. ~~Rot Overhead cam can still stutter when mouse is moving super fast. My guess is that its because we change local state of camera controller and it's slightly different than actual camera (due to pending smoothing that is being applied). Should we use state of camera instead (like it was before)? Edit: not relevant to tackle it here I think. ~~ FIXED

Fixes #16

### CODE REVIEW

File: rts/System/float3.h [hunk 1]
```diff
@@ -392,6 +392,18 @@ class float3
...
+    /**
+     * @brief directional difference in rad
+     * @param other float3 to compute against
+     * @return float3 angle difference
+     *
+     * Calculates the angle between this float3
+     * and another float3
+     */
+    float3 angleDifference(const float3& other) const {
+         return float3::fmod((other - *this) + 3.0f*math::PI, 2.0f*math::PI) - math::PI;
```
Comment (lhog) [inline rts/System/float3.h:404]: 3.0f and 2.0f look a bit arbitrary.
Perhaps look into recipes here: https://stackoverflow.com/questions/2708476/rotation-interpolation

**Comment by dzosz** [comment]
alternative and somewhat simpler implementation that relies on integer division:
https://gamedev.net/forums/topic/409881-shortest-distance-in-degrees/3728521/
`int diff = t - c
diff - (diff/ 180 * 360)`

File: rts/System/SpringMath.h [hunk 2]
```diff
@@ -111,6 +111,8 @@ template<> inline float3 argmax(const float3 v1, const float3 v2) { return float
 
 // template<class T> T mix(const T v1, const T v2, const float a) { return (v1 * (1.0f - a) + v2 * a); }
 template<class T, typename T2> constexpr T mix(const T v1, const T v2, const T2 a) { return (v1 + (v2 - v1) * a); }
+template <class T, class T2> constexpr T mixRotation(const T v1, const T v2, const T2 a) { return v1 + v1.angleDifference(v2)*a; }
```
Comment (lhog) [inline rts/System/SpringMath.h:114]: if `angleDifference` is called in a templated context then you probably want to move `angleDifference` here to SpringMath too and make it work with float and float3

File: rts/Game/Camera/RotOverheadController.cpp [hunk 3]
```diff
@@ -44,10 +46,11 @@ void CRotOverheadController::KeyMove(float3 move)
 
 void CRotOverheadController::MouseMove(float3 move)
 {
-	camera->SetRotY(camera->GetRot().y + mouseScale * move.x);
-	camera->SetRotX(camera->GetRot().x + mouseScale * move.y * move.z);
-	camera->SetRotX(Clamp(camera->GetRot().x, math::PI * 0.4999f, math::PI * 0.9999f));
-	dir = camera->GetDir();
+	auto rot = CCamera::GetRotFromDir(dir);
```
Comment (lhog) [inline rts/Game/Camera/RotOverheadController.cpp:49]: Is this called on purpose instead of `camera->GetRot()`, if so can you comment the rationale?

**Comment by lhog** [comment]
Left a few minor notes.
I assume it was tested and it fixed the issue, right?

**Comment by dzosz** [comment]
> Left a few minor notes. I assume it was tested and it fixed the issue, right?

I will provide additional context to changes. I thought that "angleDifference" might already exist somewhere in the engine code but I didn't find it. On the other side I think I saw commander making 270' turns instead of 90 when switching to building different objects.

**Comment by lhog** [comment]
There's https://github.com/beyond-all-reason/spring/blob/BAR105/rts/Sim/Units/Scripts/UnitScript.cpp#L127-L145 probably responsible for that, but as you see the implementation is done ad hoc.

File: rts/System/SpringMath.h (hunk 2 above)
Comment (dzosz) [reply rts/System/SpringMath.h]: done

**Comment by dzosz** [comment]
> 

thanks! this implementation does not require calling fmod

**Comment by dzosz** [comment]
* reimplemented mixRotation using same formula as https://github.com/beyond-all-reason/spring/blob/BAR105/rts/Sim/Units/Scripts/UnitScript.cpp#L127.
* Moved helper functions to SpringMath.inl
* fixed over-rotation bug (pending rotation over 180 deg would smooth in opposite direction) by hard-limiting it to 180. 
* added missing ClampRad function. unfortunately it takes range (0, 2pi) instead (-pi, pi)

**Comment by lhog** [comment]
Awesome, can you untouch BARb revision change you did here: https://github.com/beyond-all-reason/spring/pull/335/commits/ab6aaddf224d5379d2d01b7866a6952ea0d8dcb7 (most probably added by mistake)

File: rts/System/SpringMath.h [hunk 4]
```diff
@@ -154,6 +160,28 @@ float ClampRad(float f) _const _warn_unused_result;
...
+ */
+float GetRadAngleToward(float f1, float f2);
+
+
+/**
+ * @brief Shortest angle in radians for float3
+ * @param v1 float3 first compare value
+ * @param v2 float3 second compare value
+ */
+float GetRadAngleToward(float v1, float v2);
```
Comment (dzosz) [inline rts/System/SpringMath.h:182]: typo. let me fix that

**Comment by lhog** [comment]
Thanks. Is it ready for merge?

**Comment by dzosz** [comment]
> Thanks. Is it ready for merge?

Yes it is.
Thanks!

**Pull request merged**